service/matching/user_data_manager.go

```diff
@@ -575,7 +584,28 @@ func (m *userDataManagerImpl) HandleGetUserDataRequest(
 		} else if err != nil {
 			return nil, err
 		}
-		if req.WaitNewData && userData.GetVersion() == version {
+		if userData.GetVersion() > version {
+			resp.UserData = userData
+			m.logger.Info("returning user data",
+				tag.NewBoolTag("long-poll", req.WaitNewData),
+				tag.NewInt64("request-known-version", version),
+				tag.UserDataVersion(userData.Version),
+			)
+		} else if userData != nil && userData.Version < version && m.store != nil {
+			// When m.store == nil it means this is a non-owner partition, so it is possible
+			// for the requested version to be greater than the known version if there are
+			// concurrent user data updates in flight. We do not log an error in that case.
+
+			// This is highly unlikely to happen in the owner/root partition but may happen
+			// due to an edge case in during ownership transfer.
+			// We rely on client retries in this case to let the system eventually self-heal.
+			m.logger.Error("requested task queue user data for version greater than known version",
+				tag.NewInt64("request-known-version", version),
+				tag.UserDataVersion(userData.Version),
+			)
+			return nil, errRequestedVersionTooLarge
```

Comment:
Previously, when a non-root partition had out-of-date user data, we returned `errRequestedVersionTooLarge`. Now we only return that if the root partition is out of date, so the non-root partition won't be "self-healed" through client retries.

I just want to make sure that the non-root partition does not return success to the caller until it actually has the up-to-date user data.

It looks like on line 608, partitions with out-of-date versions trigger a `wait on userDataChanged channel`, and if they receive that signal, they'll continue through the for-loop and possibly succeed this time -- is that how non-root partitions are now expected to retry and eventually get the correct data and return success?

Reply:
yes, previously we'd error if the known version was higher than the loaded one, now we keep fetching until we pass the known version (for child partitions).